### DISCUSSION TITLE
Using Docker to containerize services.

### DISCUSSION THREAD

**Discussion by JamesM92** [iiab/iiab#2852] [Ideas]
[Edit: I've been a fan of IIAB for many years, love the concept, and keep a PI4 with a 2tb nvme drive attached to run my own IIAB server at home. I have not been involved with the project, and lack the programming skills needed to assist with this project. So forgive me if my thoughts are out of place here but I'd like to share them.]

Was just wondering if it had been considered to use Docker to containerize the different web services inside IIAB.

I figure this would cause some over head increases on the pi's, but would make maintaining code, and allowing services to be added/removed/serviced more easily.

Also many of the services being incorporated into IIAB have official docker images that are maintained, which would help limit the amount of code maintenance needed by the IIAB group. 

Kiwix Server: https://hub.docker.com/r/kiwix/kiwix-serve
Calibre Web : https://docs.linuxserver.io/images/docker-calibre-web
Next Cloud : https://hub.docker.com/_/nextcloud/
Gitea: https://docs.gitea.io/en-us/install-with-docker/

Etc. not all of the current services have docker images they maintain but a good number of them do.

(Also if yall did go with a Docker scheme, portainer makes a great management portal for docker https://www.portainer.io/)

I'm not a proficent enough programmer to try and implement anything, but figured id give a shout out about it cause i think itd make a cool addition to IIAB.

**Comment by holta** [comment]
Certainly a more complicated (and controversial!) question than it would appear.

FYI this is also mentioned near the bottom of PR #2837.

**Comment by JamesM92** [comment]
**#2837  Hermanio** 

> There are some caveats, though:
> 
>     there is some performance overhead, which can in theory be a problem on some lower-end devices, such as Raspberry Pi. From my own testing it's not catastrophic, but on very restricted devices (Raspberry Pi 1, Orange Pi Zero with 512MB RAM) it's still noticeable.
>     for some services you would have to build and publish your own images due to preexisting ones being out of date or not trustworthy
>     there might be some networking-related changes that may or may not need to be done, as Docker does a lot of magic under the hood regarding that.

1)
I know the goal of the project is to keep IIAB functional on cheap/low-end devices as much as possible. But i do think there should be some level of phase out. That some equipment is so old or hard to come by anymore its not feasible to try and maintain it anymore. a Pi1 is harder to come across than many of the newer models and the price difference between the pi1 to pi2 or a pi3 id consider minimal. So just because its not able to run on a 4 function calculator doesn't mean its not feasible.

There is also the consideration that if there were a change to docker, it would only affect versions going forward. if for some reason it is completely unusable on a pi1, people could decide to stay on the version where it did work. Yes it wouldn't be maintained anymore, but with limited volunteers to update and maintain code, some limitations of support should be expected.

2)  
True, not every service has a reliable trustworthy and maintained image that could be readily deployed. But many services that have active communities are moving towards it. 

While if your having to go through the effort to make / maintain code for things to live inside IIAB already, making it inside a docker image that is accessible and usable for other purpose is only a bonus. 



**#2837 Holta** 


> My own opinion is that it's impossible to keep up with ever-changing distro wars.




this is one of the main selling point for docker as the images are platform/OS independent, allowing mac's, window's, and all the flavors of linux to run the same image and code. And being immune to a incidental OS update that may have conflicts with the existing code.

(Another feature i like of docker is if a service freexes or crashes, you can kill it and restart it without having to kill the whole device to do a reboot)
